data/transactions/logic/resources.go

```diff
@@ -0,0 +1,359 @@
+	err := cx.requireHolding(hdr.Sender, tx.XferAsset)
+	if err != nil {
+		return fmt.Errorf("axfer Sender: %w", err)
+	}
+	err = cx.requireHolding(tx.AssetReceiver, tx.XferAsset)
+	if err != nil {
+		return fmt.Errorf("axfer AssetReceiver: %w", err)
+	}
+	err = cx.requireHolding(tx.AssetSender, tx.XferAsset)
+	if err != nil {
+		return fmt.Errorf("axfer AssetSender: %w", err)
+	}
+	err = cx.requireHolding(tx.AssetCloseTo, tx.XferAsset)
+	if err != nil {
+		return fmt.Errorf("axfer AssetCloseTo: %w", err)
+	}
+	return nil
+}
+
+func (r *resources) fillAssetFreeze(hdr *transactions.Header, tx *transactions.AssetFreezeTxnFields) {
+	r.sharedAccounts[hdr.Sender] = struct{}{}
+	id := tx.FreezeAsset
+	r.sharedAsas[id] = struct{}{}
+	r.shareAccountAndHolding(tx.FreezeAccount, id)
+}
+
+func (cx *EvalContext) allowsAssetFreeze(hdr *transactions.Header, tx *transactions.AssetFreezeTxnFields) error {
+	err := cx.requireHolding(tx.FreezeAccount, tx.FreezeAsset)
+	if err != nil {
+		return fmt.Errorf("afrz FreezeAccount: %w", err)
+	}
+	return nil
+}
+
+func (r *resources) fillApplicationCall(ep *EvalParams, hdr *transactions.Header, tx *transactions.ApplicationCallTxnFields) {
+	txAccounts := make([]basics.Address, 0, 2+len(tx.Accounts)+len(tx.ForeignApps))
+	txAccounts = append(txAccounts, hdr.Sender)
+	txAccounts = append(txAccounts, tx.Accounts...)
+	for _, id := range tx.ForeignAssets {
+		r.sharedAsas[id] = struct{}{}
+	}
+	// Make the app account associated with app calls available. We
+	// don't have to add code to make the accounts of freshly created
+	// apps available, because that is already handled by looking at
+	// `createdApps`.
+	if id := tx.ApplicationID; id != 0 {
+		txAccounts = append(txAccounts, ep.getApplicationAddress(id))
+		r.sharedApps[id] = struct{}{}
+	}
+	for _, id := range tx.ForeignApps {
+		txAccounts = append(txAccounts, ep.getApplicationAddress(id))
+		r.sharedApps[id] = struct{}{}
+	}
+	for _, address := range txAccounts {
+		r.sharedAccounts[address] = struct{}{}
+
+		for _, id := range tx.ForeignAssets {
+			r.shareHolding(address, id)
+		}
+		// Similar to note about app accounts, availableLocals allows
+		// all createdApps holdings, so we don't care if id == 0 here.
+		if id := tx.ApplicationID; id != 0 {
+			r.shareLocal(address, id)
+		}
+		for _, id := range tx.ForeignApps {
+			r.shareLocal(address, id)
+		}
+	}
+
+	for _, br := range tx.Boxes {
+		var app basics.AppIndex
+		if br.Index == 0 {
+			// "current app": Ignore if this is a create, else use ApplicationID
+			if tx.ApplicationID == 0 {
+				// When the create actually happens, and we learn the appID, we'll add it.
+				continue
+			}
+			app = tx.ApplicationID
+		} else {
+			// Bounds check will already have been done by
+			// WellFormed. For testing purposes, it's better to panic
+			// now than after returning a nil.
+			app = tx.ForeignApps[br.Index-1] // shift for the 0=this convention
+		}
+		r.boxes[boxRef{app, string(br.Name)}] = false
+	}
+}
+
```

Comment:
I guess I don't understand why there aren't any checks here.

Reply:
What kind of checks do you expect?

PS, it often seems like I'm seeing a different highlight than, perhaps, you intended.  I'm seeing lines 314-317, which have very little going on.

Reply:
A few lines up, there's a comment about not doing a bounds check, so that's probably what you are talking about.

There's no need to check because the code in transaction.WellFormed has this:
```
	for i, br := range tx.Boxes {
		// recall 0 is the current app so indexes are shifted, thus test is for greater than, not gte.
		if br.Index > uint64(len(tx.ForeignApps)) {
			return fmt.Errorf("tx.Boxes[%d].Index is %d. Exceeds len(tx.ForeignApps)", i, br.Index)
		}
	}
```
so at this point, we know `br.Index` is low enough.

Reply:
ah.  yes. I placed this comment specifically at line 320 through github.dev interface, "return nil" I'll have to take a closer look at where I click.
The return nil when the caller and callee are both, say, the resource sharing version.  In allowsAssetTransfer there are require* checks but here there are none.

Reply:
oh, ok, here's the thinking.

If the called application is >= resourceSharingVersion, then it's going to do all of its access checks using the cx.available structures - it won't access anything it shouldn't.

In the other direction, if the _calling_ application is before resourceSharingVersion, then it can only place items in the foreign arrays that it has access to (using the old rules).  They are more stringent, there's no way it can supply access to anything the callee shouldn't see. (Think of it this way: before resourceSharingVersion, nothing like this check even existed.)

We are only worried about the case that the caller has access to each half of a cross-product resource, but not the cross-product itself.  That happens if a resourceSharingVersion program gets access to an account through some other transactions and, say, an asset through another transaction.  That program does not have access to the holding for that account and asset. But, if it places those things in the foreign arrays of the program it's about to call, and that program is old (so it uses the old rules) it will simply look at the foreign arrays and conclude it has access.  So the called app has access to something the caller did not.  Neither should have such access.

This is very subtle, and I'm very happy to think about weird cases that this does not handle correctly.  But I think this does it.

Reply:
i see so allowsAssetTransfer has some of these cross-product checks (holdings checks) to prevent accidental cross-product availability, but appcall has none of those.  except that things get weird when the caller is old but the callee is new (which I guess would even only happen if the callee had its program contents updated)